src/rpc/rpc.cc

```diff
@@ -1185,6 +1185,14 @@ future<> server::connection::send_unknown_verb_reply(std::optional<rpc_clock_typ
       ).discard_result();
   }
 
+  future<> server::drop_connection(connection_id id) {
```

Comment:
Why can't it be void? In fact doesn't it has to be void since if you want to drop the connection from inside the handler it may deadlock otherwise since connection dropping will wait for all rpc handler to complete while the handler will wait for connection to stop?

Reply:
> Why can't it be void?

It can but `future<>` inteface is more powerful. If the user wants `void` they can do `(void)server.drop_connection()`. If they want to wait they can wait on the future.
If I used a `void` interface they wouldn't have the possibility to wait.

For the case of node banning in Scylla if we add a node to `left` and perform a global read barrier, and the barrier finishes, this should guarantee that the node is globally banned. So we want to wait on each node when it's dropping the connections before returning from the barrier.

> In fact doesn't it has to be void since if you want to drop the connection from inside the handler it may deadlock otherwise since connection dropping will wait for all rpc handler to complete while the handler will wait for connection to stop?

Does connection dropping wait for RPC handlers to finish? Because the test that I added is not deadlocking, and I'm calling `stop()` and waiting from inside the handler. It's a two-way RPC handler even.

Reply:
> > Why can't it be void?
> 
> It can but `future<>` inteface is more powerful. If the user wants `void` they can do `(void)server.drop_connection()`. If they want to wait they can wait on the future. If I used a `void` interface they wouldn't have the possibility to wait.

User cannot just do `(void)` since it has to wait for the future somehow. It may move it to a background fiber but this is just a hassle in case the wait is not needed.

> 
> For the case of node banning in Scylla if we add a node to `left` and perform a global read barrier, and the barrier finishes, this should guarantee that the node is globally banned. So we want to wait on each node when it's dropping the connections before returning from the barrier.
> 
> > In fact doesn't it has to be void since if you want to drop the connection from inside the handler it may deadlock otherwise since connection dropping will wait for all rpc handler to complete while the handler will wait for connection to stop?
> 
> Does connection dropping wait for RPC handlers to finish? Because the test that I added is not deadlocking, and I'm calling `stop()` and waiting from inside the handler. It's a two-way RPC handler even.

There is a reply gate that is waited on in `server::stop()`, so not sure how your test works. Anyway allowing it just asking for a disaster to happen.

Reply:
In fact it just shows that getting the connection stop() future resolved does not guaranty that there are no handlers from the node we try to ban running.

Reply:
I think we need to move server gate into the connections and have `future<> stop()` that aborts the connection (by calling `void abort()` and waits for its gate and call  `abort()` in CLIENT_ID without waiting. There is nothing to wait there for. This is in the case we really want to make sure no handler from the banned node are running after `ban_node()` completes.

Reply:
> > User cannot just do (void) since it has to wait for the future somehow.
> 
> Why?
Because all features has to be waited. Sometime it may be done indirectly by having fibers collected in the gate. But you still need to wait for them at some point.

> 
> Discarding the future will be equivalent to a `void` interface.
You need to collect them on server::stop or something. Why? That's the rule. We do not want to have unattended fibers running during shutdown. They may be scheduled when we do not expect. 

> 
> > In fact it just shows that getting the connection stop() future resolved does not guaranty that there are no handlers from the node we try to ban running.
> 
> That's fine, we don't need to guarantee that handlers are not running. But any responses from those handlers or any later handler calls will not reach the node after `ban_node()` completes.

Why do we care that a response will not get a node we are considered banned will not reach it? For all we know the node may burn from this point. It looks to me that if we do want anything at all we want to not execute any rpcs from the banned node after future resolves locally.


> 
> > I think we need to move server gate into the connections and have future<> stop() that aborts the connection (by calling void abort() and waits for its gate and call abort() in CLIENT_ID without waiting. There is nothing to wait there for. This is in the case we really want to make sure no handler from the banned node are running after ban_node() completes.
> 
> I'm not sure I understand your point.
> 
> Do you suggest having `void server::drop_connection(id)`, which calls `abort()` and returns immediately?

I suggest having 'future<> server::drop_connection(id)` which actually guaranties that no handlers are running any longer and use it in `ban_host()` (if we really need the guaranty). And call existing `void connection::abort()` in `CLIENT_INFO` verb.

> 
> IIUC I won't even be able to test this consistently, the test that I wrote will have to be modified like this:
> 
> ```c++
> SEASTAR_TEST_CASE(test_rpc_drop_connection) {
>     return rpc_test_env<>::do_with_thread(rpc_test_config(), [] (rpc_test_env<>& env) {
>         test_rpc_proto::client c1(env.proto(), {}, env.make_socket(), ipv4_addr());
>         env.register_handler(1, [] (rpc::client_info& cinfo, int x) {
>             if (x) {
>                 cinfo.server.drop_connection(cinfo.conn_id);
>             }
>             return make_ready_future<int>(0);
>         }).get();
>         auto f = env.proto().make_client<int (int)>(1);
>         BOOST_REQUIRE_EQUAL(f(c1, 0).get0(), 0);
>         BOOST_REQUIRE_THROW(f(c1, 1).get0(), rpc::closed_error);
>         c1.stop().get0();
>     });
> }
> ```
> 
> the test will become flaky, because the response returned after `drop_connection` may actually reach the caller after this modification, no? The `void` interface loses guarantees that the `future<>` interface gave you if you waited on it. It's just strictly worse.

I care about the interface, not weather you can come up with a simple test. I am sure we can come up with a way to test it when we agree on the interface itself.

Reply:
> Because all features has to be waited. Sometime it may be done indirectly by having fibers collected in the gate. But you still need to wait for them at some point.

> You need to collect them on server::stop or something. Why? That's the rule. We do not want to have unattended fibers running during shutdown. They may be scheduled when we do not expect.

"That's the rule" is not a reason. And not all futures have to be waited on.

The future returned from `stop()` is obtained from a shared promise. The promise is set at the end of connection process loop. There is no unattended fiber. It's fine to discard the future.

A `void` implementation would just call `abort()` and then not get the future from the shared promise. What's the difference between not getting a future from a shared promise, and getting it then discarding, other than minor performance difference? Asking theoretically, because I do actually want to wait on the future.

But for now let's focus on the actual problem, what we want the function to guarantee us.

> It looks to me that if we do want anything at all we want to not execute any rpcs from the banned node after future resolves locally.

Unfortunately we will execute further RPCs because new connections will be established each time the node tries to contact us.
But every such new connection will start by receiving CLIENT_ID RPC which will drop the connection.

I don't see a way to completely prevent RPCs from being executed.

It may be that the connection dropping in CLIENT_ID handler races with other handlers executed concurrently - can they? Or will `abort()` prevent further handlers from being executed if we call it immediately without yielding?

> I suggest having `future<> server::drop_connection(id)` which actually guaranties that no handlers are running any longer 

By using a gate? How will it differ from the currently used shared_promise solution? Will the gate be closed at a different point? The shared_promise is set when we finish the connection process loop.

> call existing `void connection::abort()` in `CLIENT_INFO` verb.

Note that it would have to be `void server::abort_connection(id)` because Scylla won't have access to `connection` itself.

Reply:
> > Because all features has to be waited. Sometime it may be done indirectly by having fibers collected in the gate. But you still need to wait for them at some point.
> 
> > You need to collect them on server::stop or something. Why? That's the rule. We do not want to have unattended fibers running during shutdown. They may be scheduled when we do not expect.
> 
> "That's the rule" is not a reason.

That is why I elaborated further.

> And not all futures have to be waited on.

At least indirectly? All of them. We even comment on each (void) casting how we eventually wait or have a FIXME.

> 
> The future returned from `stop()` is obtained from a shared promise. The promise is set at the end of connection 
process loop. There is no unattended fiber. It's fine to discard the future.

Implementation detail that may change. How can you tell it from just the interface `future<> drop()`? Even then you need to at least handle an error or get warnings.

> 
> A `void` implementation would just call `abort()` and then not get the future from the shared promise. What's the difference between not getting a future from a shared promise, and getting it then discarding, other than minor performance difference? Asking theoretically, because I do actually want to wait on the future.
>

The interface is different. One requires you to wait in some way another is not.
 
> But for now let's focus on the actual problem, what we want the function to guarantee us.
> 
> > It looks to me that if we do want anything at all we want to not execute any rpcs from the banned node after future resolves locally.
> 
> Unfortunately we will execute further RPCs because new connections will be established each time the node tries to contact us. But every such new connection will start by receiving CLIENT_ID RPC which will drop the connection.

That's fine. We except that we can execute an RPC that will veto the connection. May be we can extend RPC negotiation to support banning...

> 
> I don't see a way to completely prevent RPCs from being executed.

If we do not want to prevent RPCs from been executed what are we trying to achieve anyway?

> 
> It may be that the connection dropping in CLIENT_ID handler races with other handlers executed concurrently - can they? Or will `abort()` prevent further handlers from being executed if we call it immediately without yielding?

It will. In fact we relay on it because `CLIENT_ID` populates info other RPC relies on.

> 
> > I suggest having `future<> server::drop_connection(id)` which actually guaranties that no handlers are running any longer
> 
> By using a gate?

Yes, by moving server gate into each connection and wait for all of them in `server::stop()`. Again only if we actually convinced we need the guaranty!

> How will it differ from the currently used shared_promise solution? Will the gate be closed at a different point? The shared_promise is set when we finish the connection process loop.

Yes, the gate will be closed when each currently running handler exits. shared promise will be resolved when connection is fully closed. I do not think connection closing waits for all handler so return since otherwise your test would have hanged. I do not see in the code either.

> 
> > call existing `void connection::abort()` in `CLIENT_INFO` verb.
> 
> Note that it would have to be `void server::abort_connection(id)` because Scylla won't have access to `connection` itself.

Right.

Reply:
> Implementation detail that may change. How can you tell it from just the interface future<> drop()? Even then you need to at least handle an error or get warnings.

Easy, I could write a comment at the declaration: "It's safe to discard the future if you won't need to wait."
But nevermind that

> One requires you to wait in some way

Correction: makes it possible to wait
But nevermind that

> It will. In fact we relay on it because CLIENT_ID populates info other RPC relies on.

Good

> Yes, by moving server gate into each connection and wait for all of them in server::stop(). Again only if we actually convinced we need the guaranty!

Now I don't think we do

So let's just do it with `abort()` and finish the discussion, `abort()` is good enough

Reply:
I'm not sure though what test should I write now

Reply:
What about sending another RPC after the first one and see if it success?

Reply:
> What about sending another RPC after the first one and see if it success?

Thought about it but the second RPC will just happen on a newly established connection, right?
Although maybe if we don't yield when sending the two in succession...
It looks a bit "slimy" but should work

Reply:
> > What about sending another RPC after the first one and see if it success?
> 
> Thought about it but the second RPC will just happen on a newly established connection, right? Although maybe if we don't yield when sending the two in succession... It looks a bit "slimy" but should work

Re-connection magic happens in Scylla on messaging_service level, not in the RPC (this is problematic in itself, but this is how it is). In the test if you use the same client you use the same connection.